kernel/src/osquery.cpp

```diff
@@ -0,0 +1,259 @@
+
+      if (osquery.buffer != NULL) {
+        // We don't want to allocate a second buffer.
+        err = -EINVAL;
+        goto error_exit;
+      }
+
+      if ((err = allocate_user_kernel_buffer(alloc->size, &(alloc->buffer)))) {
+        goto error_exit;
+      }
+
+      dbg_printf("IOCTL alloc: size %lu, location %p\n",
+                 alloc->size, alloc->buffer);
+      break;
+    default:
+      err = -ENOTTY;
+      goto error_exit;
+      break;
+  }
+  return 0;
+error_exit:
+  return err;
+}
+
+// OSQuery character device switch structure.
+static struct cdevsw osquery_cdevsw = {
+    osquery_open,    // open_close_fcn_t *d_open;
+    osquery_close,   // open_close_fcn_t *d_close;
+    eno_rdwrt,       // read_write_fcn_t *d_read;
+    eno_rdwrt,       // read_write_fcn_t *d_write;
+    &osquery_ioctl,  // ioctl_fcn_t      *d_ioctl;
+    eno_stop,        // stop_fcn_t       *d_stop;
+    eno_reset,       // reset_fcn_t      *d_reset;
+    NULL,            // struct tty      **d_ttys;
+    eno_select,      // select_fcn_t     *d_select;
+    eno_mmap,        // mmap_fcn_t       *d_mmap;
+    eno_strat,       // strategy_fcn_t   *d_strategy;
+    eno_getc,        // getc_fcn_t       *d_getc;
+    eno_putc,        // putc_fcn_t       *d_putc;
+    0                // int               d_type;
+};
+
+kern_return_t OSQueryStart(kmod_info_t *ki, void *d) {
+  dbg_printf("OSQuery kernel module starting!\n");
+  asm("int $3");
+
+
+  osquery.major_number = cdevsw_add(osquery.major_number, &osquery_cdevsw);
+  if (osquery.major_number < 0) {
+    dbg_printf("Could not get a major number!\n");
+    goto error_exit;
+  }
+
+  osquery.devfs = devfs_make_node(makedev(osquery.major_number, 0),
+                                          DEVFS_CHAR, UID_ROOT, GID_WHEEL,
+                                          0644, "osquery", 0);
+  if (osquery.devfs == NULL) {
+    dbg_printf("Could not get a devfs entry!\n");
+    goto error_exit;
+  }
+
+  return KERN_SUCCESS;
+error_exit:
+  if (osquery.devfs != NULL) {
+    devfs_remove(osquery.devfs);
+    osquery.devfs = NULL;
+  }
+
+  if (!(osquery.major_number < 0)) {
+    if (cdevsw_remove(osquery.major_number, &osquery_cdevsw) < 0) {
+      panic("OSQuery kext:  Cannot remove osquery from cdevsw");
+    }
+  }
+  return KERN_FAILURE;
+}
+
+kern_return_t OSQueryStop(kmod_info_t *ki, void *d) {
+  dbg_printf("OSQuery kernel module stoping!\n");
+
+  if (osquery.buffer) {
+    IOFreeAligned(osquery.buffer, osquery.buf_size);
+    osquery.buffer = NULL;
+  }
+
+  devfs_remove(osquery.devfs);
+  osquery.devfs = NULL;
+
+  if (cdevsw_remove(osquery.major_number, &osquery_cdevsw) < 0) {
+    panic("OSQuery kext:  Cannot remove osquery from cdevsw");
+  }
+
+  return KERN_SUCCESS;
+}
+
+extern "C" {
+extern kern_return_t _start(kmod_info_t *ki, void *data);
+extern kern_return_t _stop(kmod_info_t *ki, void *data);
+}
+
+KMOD_EXPLICIT_DECL(com.facebook.kern.OSQuery, "1.0.0d1", _start, _stop)
```

Comment:
Let's align the version with our build version. Check out the root Makefile and how it [finds the version](https://github.com/facebook/osquery/blob/master/Makefile#L2). You can pass that as a define maybe?